doc/deghosting_mask.1
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "DEGHOSTING_MASK ""1"""
.TH DEGHOSTING_MASK "1" "2010-12-26" """Version: 2010.5.0""" "HUGIN"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
deghosting_mask \- Creates mask for removing ghosting in images
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBdeghosting_mask\fR [options] \fI<image1\fR> \fI<image2\fR> \fI...\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Takes a series of aligned photos, usually three or more, and generates masks
for removal of ghosts \- Data that only appears in a minority of photos.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-o|\-\-output\fR <path>" 4
.IX Item "-o|--output <path>"
Prefix for output masks
.IP "\fB\-s|\-\-sigma\fR <float>" 4
.IX Item "-s|--sigma <float>"
Standard deviation of Gaussian weighting function (\s-1SIGMA\s0 > 0); default: 30
.IP "\fB\-i|\-\-iterations\fR <int>" 4
.IX Item "-i|--iterations <int>"
Number of iterations, default is (\s-1ITER\s0 > 0) default: 4
.IP "\fB\-t|\-\-threshold\fR <float>" 4
.IX Item "-t|--threshold <float>"
Threshold; default: 150
.IP "\fB\-c|\-\-contrast\fR <float>" 4
.IX Item "-c|--contrast <float>"
Change constrast before applying threshold; default: 1.3
.IP "\fB\-a|\-\-advanced\fR <f|g|m|t|w>" 4
.IX Item "-a|--advanced <f|g|m|t|w>"
Advanced settings. Possible options are:
.RS 4
.IP "\fBf\fR Use gray images for computation. It's about two times faster but it usually returns worse results.  You also have to change threshold to smaller value (around 100)" 4
.IX Item "f Use gray images for computation. It's about two times faster but it usually returns worse results.  You also have to change threshold to smaller value (around 100)"
.PD 0
.IP "\fBg\fR Use gamma 2.2 correction instead of logarithm if input images are \s-1HDR\s0" 4
.IX Item "g Use gamma 2.2 correction instead of logarithm if input images are HDR"
.IP "\fBm\fR Do not scale image, \s-1NOTE:\s0 slows down process" 4
.IX Item "m Do not scale image, NOTE: slows down process"
.IP "\fBt\fR Use simple threshold, may result in holes in images" 4
.IX Item "t Use simple threshold, may result in holes in images"
.ie n .IP "\fBw\fR Compute ""complete"" weights, not only probabilities" 4
.el .IP "\fBw\fR Compute ``complete'' weights, not only probabilities" 4
.IX Item "w Compute complete weights, not only probabilities"
.RE
.RS 4
.RE
.IP "\fB\-w|\-\-save\fR <i|w>" 4
.IX Item "-w|--save <i|w>"
.PD
Advanced save settings
.RS 4
.IP "\fBi\fR Save initial weigths" 4
.IX Item "i Save initial weigths"
.PD 0
.IP "\fBw\fR Save generated weigths" 4
.IX Item "w Save generated weigths"
.RE
.RS 4
.RE
.IP "\fB\-b\fR <int>" 4
.IX Item "-b <int>"
.PD
Image cache \s-1BLOCKSIZE\s0 in kilobytes; default: 2048KB
.IP "\fB\-m\fR <int>" 4
.IX Item "-m <int>"
Set image \s-1CACHESIZE\s0 in megabytes; default: 1024MB
.IP "\fB\-h|\-\-help\fR" 4
.IX Item "-h|--help"
Display this help
.IP "\fB\-v|\-\-verbose\fR <0|1>" 4
.IX Item "-v|--verbose <0|1>"
Verbose, repeat for more verbose output
.SH "AUTHORS"
.IX Header "AUTHORS"
LukA\*~XA\*oX JirkovskA\*~X
.SH "POD ERRORS"
.IX Header "POD ERRORS"
Hey! \fBThe above document had some coding errors, which are explained below:\fR
.IP "Around line 86:" 4
.IX Item "Around line 86:"
You forgot a '=back' before '=head1'


NEWS
Newsworthy changes in 0.7 beta 5
--------------------------------
* much improved vignetting and exposure correction,
  http://hugin.sf.net/tech
* ability to create HDR images during stitching
  (still extremly memory hungry, will be improved
   in the next version).
* fixed bad HFOV/VFOV limits of some projections
* OSX build should not crash all the time
* Added "Apply template" function, to apply the
  settings of a previously saved project to
  the current images.
* enabled "load lens data" button in dialog asking
  for focal length/HFOV.

Newsworthy changes in 0.7 beta 4
-------------------------
* update panotools to latest version
* Albers equal area conic and Miller cylindrical projection
* Support for straight line control points
* Fixed running of Autopano-SIFT integration on windows
* (linux) work around bugs in wxWidgets (fixes double deletion
  of control points when pressing the Del key)
* Force even values for width for 360 degree equirectangular
  panoramas.

Newsworthy changes in 0.7 beta 3
-------------------------
* Assistant panel, enables 3 click stitching (well, almost ;-)
* Center and rotate panorama by mouse clicks in preview window.
  Also added a numerical transform dialog
* automatic straighten
* support for Lambert and Lambert Azimuthal projections
* separate crop tab replaces buggy crop dialog
* crop is saved in lens .ini files
* Rotated display of images in control point selector, depending
  on the roll value
* allow images with and without alpha channels in the same project
* option to open final panorama in image editor after stitching
* present output of command line utils in nice window (unix)
* many small bugfixes
* removed image center dialog


Newsworthy changes in 0.6.1
---------------------------
* Autopano-SIFT integration was broken in 0.6 on Unix and OSX.
* Support for displaying HDR files within hugin, using log or gamma mapping
* .hdr file support (tif HDR files have been always useable)
* reduced memory usage, better image caching strategy in GUI.
* fixed crashes and hangs when using the preview auto update and avoid
  unnessecary remapping of images.
* vignetting correction estimation speed up.
* nona and nona_gui can be used to do template stitching, by specifying
  images on the command line (or dnd shortcut under windows)
* distribute fulla.exe with windows package (it was accidentally
  not included in the 0.6 package)
* universal build for Intel Macs, fixed hang after enblend or autopano-sift
  call
* projection can be changed in preview.
* removed all references to malicious panotools.info website
* various smaller bugfixes

Newsworthy changes in 0.6
-------------------------
* updated vigra to version 1.4. All sourcecode of hugin is now
  released under approved open source licences. This means there
  is no problem with including hugin in debian and ubuntu anymore.
* Copy ICC profiles from input images to output
* Added vignetting correction and brighness correction
  during stitching.
  Vignetting can be corrected by addition or division.
  The correction can be done with a flatfield image or an
  radial polynomial
* Functionality to estimate the vignetting polynomial
  directly from the overlapping images. This is an experimental feature.
* Multithreaded remapping. hugin and nona fully utilize multiple
  CPUs or multi core CPUs during preview generation and stitching
* support for dcraw 16 bit .ppm files
* Disabled PTOptimizer progress display on OSX. This should fix
  the crashes some people have experienced.
* New command line tool (fulla), for distortion, TCA and vignetting correction.
  Can use PTLens database (old format only).
* reorganized GUI layout.
* Czech translation added

Newsworthy changes in 0.5
-------------------------
* Added Catalan, Chinese and Hungarian translation
* Support TIFF compressions: NONE, DEFLATE, LZW
* option to create cropped tiff files for usage with enblend 2.4
* Mac OSX: fixed bug during calling PTOptimizer executable
* Windows only:
    hugin is shipped with the pano12.dll build with MSVC.
    This might lead to crashes when using PTStitcher.exe
    In that case replace pano12.dll with pano12_for_usage_with ptstitcher.dll

Newsworthy changes in 0.5 rc2
-----------------------------
* Removed all non-opensource sourcecode from CVS
* Fixed various crashes
* Japanese, Russian, Brazilian and Dutch translation
* Proper default arguments for Autopano-SIFT on Windows
  (users of hugin 0.5 r1 or less need to restore the default
   config)

Newsworthy changes in 0.5 rc1
-----------------------------
* Options on the menu and toolbar that are irrelevant
  are now disabled.
* enblend can now be called from nona even if it has 
  spaces in the path.

Newsworthy changes in 0.5 beta7
-------------------------------
* Updated translations
* Fixes to the build system on linux and windows

Newsworthy changes in 0.5 beta6
-------------------------------
* All files are now in source tarball
* French translation now complete

Newsworthy changes in 0.5 beta5
-------------------------------

* Support for modifying the size of the right/left panes in many tabs.
* All controls can now be reached on smaller displays as many areas 
  of the interface are now scrollable.
* All image control buttons in the preview tab can now be accessed, 
  allowing support for far more images.

* OSX build now supports unicode.

* Enblend arguments specified in preferences are now passed on correctly.
* Multiple tiff file outputs are no longer deleted incorrectly.
* Simultaneous scrolling in control points editor tabs respects borders properly.


src/deghosting/deghosting_mask.cpp

/**
 * Tool for creating b/w alpha masks to eliminate ghosting artifacts
 * Copyright (C) 2009  Lukáš Jirkovský <l.jirkovsky@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 *Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

#include <iostream>
#include <string>
#include <cstring>

#include <hugin_config.h>
#include <hugin_version.h>
// for stripExtension
#include <hugin_utils/utils.h>
// for exportImage
#include <vigra_ext/impexalpha.hxx>

#include "deghosting.h"
#include "support.h"
#include "threshold.h"
#include "denoise.h"

// deghosting algorithms
#include "khan.h"

#include <getopt.h>
#ifdef WIN32
    #define snprintf _snprintf
#endif

using std::cout;
using std::cerr;
using std::endl;
using std::string;

using namespace deghosting;

// options for otherFlags
static const uint16_t SAVE_GENWEIGHTS = 1;
static const uint16_t OTHER_GRAY = 2;

// globals containing settings
static int iterations = 4;
static double sigma = 30;
static uint16_t flags = ADV_ONLYP + ADV_MULTIRES;
static uint16_t otherFlags = 0;
static uint16_t otherThresholdFlags = 0;
static uint16_t debugFlags = 0;
static EMoR response(0.0f);
static int verbosity = 0;
static double thresholdLim = 150;
static double contrast = 1.3;

/** function handling advanced options
 */
void parseOptions_advanced(char* optarg) {
    for(char *c = optarg; *c; c++) {
        switch(*c) {
            case 'f':
                otherFlags += OTHER_GRAY;
                break;
            case 'g':
                flags += ADV_GAMMA;
                break;
            case 'm':
                flags -= ADV_MULTIRES;
                break;
            case 't':
                otherThresholdFlags += THRESHOLD_DONTCARE;
                break;
            case 'w':
                flags -= ADV_ONLYP;
                break;
            default:
                cerr<< "Error: unknown option" << endl;
                exit(1);
        }
    }
}

/** function handling save options
 */
void parseOptions_save(char* optarg) {
    for(char *c = optarg; *c; c++) {
        switch(*c) {
            case 'i':
                debugFlags += SAVE_INITWEIGHTS;
                break;
            case 'w':
                otherFlags += SAVE_GENWEIGHTS;
                break;
            default:
                cerr<< "Error: unknown option" << endl;
                exit(1);
        }
    }
}

static void usage()
{
    cerr << "deghosting_mask: creates mask for removing ghosting in images" << endl
         << "deghosting_mask version " << DISPLAY_VERSION << endl
         << endl
         << "Usage: deghosting_mask [options] inputfile(s) " << endl
         << "   option are: " << endl
         << "     -o, --output=PREFIX       prefix for output masks" << endl
         << "     -i, --iterations=ITER     number of iterations, default is (ITER > 0)" << endl
         << "                               default: " << iterations << endl
         << "     -s, --sigma=SIGMA         standard deviation of Gaussian weighting" << endl
         << "                               function (SIGMA > 0); default: " << sigma << endl
         //<< "     -r, --response=E:M:o:R    use camera response specified in EMoR format" << endl
         << "     -t, --threshold=THRESH    threshold; default: " << thresholdLim << endl
         << "     -c, --contrast=CONTR      change constrast before applying threshold;" << endl
         << "                               default: " << contrast << endl
         << "     -a, --advanced=SET        advanced settings. Possible options are:" << endl
         << "                               f   use gray images for computation. It's about two times faster" << endl
         << "                                   but it usually returns worse results." << endl
         << "                                   You also have to change threshold to smaller value (around 100)" << endl
         << "                               g   use gamma 2.2 correction instead of logarithm if input images are HDR" << endl
         << "                               m   do not scale image, NOTE: slows down process" << endl
         << "                               t   use simple threshold, may result in holes in images" << endl
         << "                               w   compute \"complete\" weights, not only probabilities" << endl
         << "     -w, --save=SET            advanced save settings" << endl
         << "                               i   save initial weigths" << endl
         << "                               w   save generated weigths" << endl
         << "     -b BLOCKSIZE              image cache BLOCKSIZE in kilobytes; default: " <<
            (CachedFileImageDirector::v().getBlockSize() / 1024LL) << "KB" << endl
         << "     -m CACHESIZE              set image CACHESIZE in megabytes; default: " << 
            (CachedFileImageDirector::v().getAllocation() / 1048576LL) << "MB" << endl
         << "     -h, --help                display this help" << endl
         << "     -v, --verbose             verbose, repeat for more verbose output" << endl;
}

int main(int argc, char *argv[]) {
    try{
    const char * optstring = "o:i:s:r:t:c:a:w:b:m:hv";
    opterr = 0;
    int c;
    
    string outputPrefix = "weight";
    
    enum optionArgumentKind {
        NoArgument,
        StringArgument,
        DoubleArgument,
        IntegerArgument,
        ArrayArgument
    };
    
    enum optionId {
        outputID,
        iterationsID,
        sigmaID,
        responseID,
        thresholdID,
        constrastID,
        advancedID,
        saveID,
        helpID,
        verboseID
    };
      
    static struct option longOptions[] = {
        {"output", 1, 0, StringArgument},
        {"iterations", 1, 0, IntegerArgument},
        {"sigma", 1, 0, DoubleArgument},
        {"response", 1, 0, ArrayArgument},
        {"threshold", 1, 0, DoubleArgument},
        {"contrast", 1, 0, DoubleArgument},
        {"advanced", 1, 0, StringArgument},
        {"save", 1, 0, StringArgument},
        {"help", 0, 0, NoArgument},
        {"verbose", 0, 0, NoArgument},
        {0, 0, 0, 0}
    };
    
    // TEST
    // response for testing
    response.resize(5);
    response[0] = -3.59f;
    response[1] = -0.93f;
    response[2] =  0.11f;
    response[3] = -0.22f;
    response[4] =  0.34f;
       
    int optionIndex = 0;
    
    while ((c = getopt_long(argc, argv, optstring, longOptions, &optionIndex)) != -1) {
        switch (c) {
            case NoArgument: {
                if (longOptions[optionIndex].flag != 0) break;
                switch (optionIndex) {
                    case helpID:
                        usage();
                        return 0;
                    case verboseID:
                        verbosity++;
                        break;
                    default:
                        cerr << "There's a problem with parsing options" << endl;
                        return 1;
                }
                break;
            }
            
            case StringArgument: {
                if (longOptions[optionIndex].flag != 0) break;
                switch (optionIndex) {
                    case outputID:
                        outputPrefix = optarg;
                        break;
                    case advancedID:
                        parseOptions_advanced(optarg);
                        break;
                    case saveID:
                        parseOptions_save(optarg);
                        break;
                    default:
                        cerr << "There's a problem with parsing options" << endl;
                        return 1;
                }
                break;
            }
            
            case IntegerArgument: {
                if (longOptions[optionIndex].flag != 0) break;
                switch (optionIndex) {
                    case iterationsID:
                        iterations = atoi(optarg);
                        break;
                    default:
                        cerr << "There's a problem with parsing options" << endl;
                        return 1;
                }
                break;
            }
            
            case DoubleArgument: {
                if (longOptions[optionIndex].flag != 0) break;
                switch (optionIndex) {
                    case sigmaID:
                        sigma = atof(optarg);
                        break;
                    case thresholdID:
                        thresholdLim = atof(optarg);
                        break;
                    case constrastID:
                        contrast = atof(optarg);
                        break;
                }
                break;
            }
            
            case ArrayArgument: {
                if (longOptions[optionIndex].flag != 0) break;
                switch (optionIndex) {
                    case responseID:
                        // TODO
                        break;
                }
                break;
            }
            
            case 'o':
                outputPrefix = optarg;
                break;
            case 'i':
                iterations = atoi(optarg);
                break;
            case 's':
                sigma = atof(optarg);
                break;
            case 'r':
                // TODO
                break;
            case 't':
                thresholdLim = atof(optarg);
                break;
            case 'c':
                contrast = atof(optarg);
                break;
            case 'a':
                parseOptions_advanced(optarg);
                break;
            case 'w':
                parseOptions_save(optarg);
                break;
            case 'b': {
                const int kilobytes = atoi(optarg);
                if (kilobytes < 1) {
                    cerr << "cache block size must be 1 or more." << endl;
                    return 1;
                }
                CachedFileImageDirector::v().setBlockSize(static_cast<long long>(kilobytes) << 10);
                break;
            }
            case 'm': {
                const int megabytes = atoi(optarg);
                if (megabytes < 1) {
                    cerr << "memory limit must be 1 or more." << endl;
                    return 1;
                }
                CachedFileImageDirector::v().setAllocation(static_cast<long long>(megabytes) << 20);
                break;
            }
            case 'h':
                usage();
                return 0;
            case 'v':
                verbosity++;
                break;
        }
    }
    
    cout << endl;
    
    unsigned nFiles = argc - optind;
    if (nFiles == 0) {
        cerr << std::endl << "Error: at least three input images needed" << std::endl <<std::endl;
        usage();
        return 1;
    } else if (nFiles < 3) {
        std::cout << std::endl << "Error: You have to specify at least three images." << std::endl;
        return 1;
    }
    
    // load all images
    vector<string> inputFiles;
    for (size_t i=optind; i < (size_t)argc; i++)
    {
        inputFiles.push_back(argv[i]);
    }
    
    Deghosting* deghoster = NULL;
    
        vector<FImagePtr> weights;
        if (otherFlags & OTHER_GRAY) {
            Khan<float> khanDeghoster(inputFiles, flags, debugFlags, iterations, sigma, verbosity);
            deghoster = &khanDeghoster;
            weights = deghoster->createWeightMasks();
        } else {
            Khan<RGBValue<float> > khanDeghoster(inputFiles, flags, debugFlags, iterations, sigma, verbosity);
            deghoster = &khanDeghoster;
            weights = deghoster->createWeightMasks();
        }
        
        //deghoster->setCameraResponse(response);
                
        // save weights
        if (otherFlags & SAVE_GENWEIGHTS) {
            for (unsigned int i=0; i<weights.size(); ++i) {
                char tmpfn[100];
                snprintf(tmpfn, 99, "%s_%d.tif", outputPrefix.c_str(), i);
                ImageExportInfo exWeights(tmpfn);
                exportImage(srcImageRange(*weights[i]), exWeights.setPixelType("UINT8"));
            }
        }
        
        // apply contrast functor
        for (unsigned int i=0; i < weights.size(); ++i) {
            transformImage(srcImageRange(*(weights[i])), destImage(*(weights[i])), BrightnessContrastFunctor<FImage::PixelType>(1, contrast, 0, 255));
        }
        
        vector<BImagePtr> thresholded = threshold(weights, thresholdLim, otherThresholdFlags);
        
        // save masks with treshold applied
        for (unsigned int i=0; i<weights.size(); ++i) {
            char tmpfn[100];
            string fileName = hugin_utils::stripExtension(inputFiles[i]);
            fileName = hugin_utils::stripPath(fileName);
            snprintf(tmpfn, 99, "%s_mask.tif", fileName.c_str());
            ImageExportInfo exWeights(tmpfn);
            BImage outImg = BImage((*thresholded[i]).size());
            simpleDenoise(srcImageRange(*thresholded[i]), destImage(outImg));
            exportImage(srcImageRange(outImg), exWeights.setPixelType("UINT8"));
        }
    } catch (const std::exception & e) {
        std::cerr << "caught exception: " << e.what() << std::endl;
        return 1;
    } catch (...) {
        std::cerr << "caught unknown exception" << std::endl;
        return 1;
    }
    return 0;
}
